example/extensions/lib_custom_op/README.md

```diff
@@ -146,21 +186,160 @@ Let’s take a closer look at those registry functions:
+
+1. Run `make relu_lib`. The Makefile will invoke `NVCC` compiler to compile the CUDA kernel along with regular custom operator functions from `relu_lib.cu` to generate `librelu_lib.so` library.
+2. Run `python test_relu.py`. It’ll register the GPU `ReLU` operator in the MXNet backend, then invoke the operator by feeding an `NDArray` input with GPU context, and output the result tensor with GPU context.
+
+### Writing A Regular GPU Custom Operator
+
+Since most of the building blocks for registering GPU custom operators are the exactly same as CPU ones, the registering function for an operator supporting both GPU and CPU will look like this:
+
+```c++
+    REGISTER_OP(my_op_gpu)
+    ...
+    .setForward(forwardCPU, "cpu")
+    .setForward(forwardGPU, "gpu")
+    .setBackward(backwardCPU, "cpu")
+    .setBackward(backwardGPU, "gpu");
+```
+
+Note that you can write only the GPU funcitons, as operators don’t have to support both CPU and GPU functions (can be GPU only).
+
+After you register forward or backward functions with context name `“gpu”`, MXNet will dispatch forward or backward GPU functions you just registered when the operator is invoked with GPU context.
+
+In the registered `forwardGPU` function, specify the grid and block size and launch the CUDA kernel. In the GPU operators input and output tensors are pre-allocated on the GPU memory, just like in the CPU operator tensors are pre-allocated on the CPU memory. As a result, you don’t need to call `cudaMemcpy` to move the tensor data to GPU device.
+
+```c++
+    MXReturnValue forwardGPU(std::map<std::string, std::string> attrs,
+                             std::vector<MXTensor> inputs,
+                             std::vector<MXTensor> outputs,
+                             OpResource res) {
+        float* in_data = inputs[0].data<float>();
+        float* out_data = outputs[0].data<float>();
+        mx_stream_t cuda_stream = res.get_cuda_stream();
+        int64_t N = inputs[0].size();
+        int block = 256;
+        int grid = (N + (block - 1)) / block;
+        my_op_gpu_forward<<<grid,block,0,cuda_stream>>>(out_data, in_data, N);
+        return MX_SUCCESS;
+    }
+```
+
+Note that the `cuda_stream` object used for launching kernels is passed from MXNet backend via `OpResource` object. See below for details of `Operator Resource`.
+
+At this point all the attribute functions for each operator (`parseAttrs`, `inferShape`, etc.) run on the CPU, including the `forwardGPU` function. The only part that will actually run on the GPU is the CUDA kernel function that is launched from `forwardGPU` function (ie. my_op_gpu_forward).
+
+```c++
+    __global__ void my_op_gpu_forward(float *out, float *in, int64_t N) {
+        int tid = blockIdx.x * blockDim.x + threadIdx.x;
```

Comment:
I have a question about the balance between code readability and correctness.

In CUDA code, the number of block and thread is limited. This code works for small N, but it doesn't work when `N` is larger than the maximum number of block and thread. However, it is good to read for users.

The correct code shows as below, but it is unfriendly to read for newbies.
```c++
for (int64_t i = blockIdx.x * blockDim.x + threadIdx.x; i < N; i += blockDim.x * gridDim.x) {
  // code your computational logic here
}
```

I'm not sure which one is better.

Reply:
I agree with @wkcn, lets just replace the function body with:
```
// code your computational logic here
```
Then it can be generic and not confusing to advanced users. The function body here doesnt really give any additional meaning

Reply:
agree, let's be generic enough for this

Reply:
It is correct and readable to replace `int tid = blockIdx.x * blockDim.x + threadIdx.x;`  with the `CUDA_KERNEL_LOOP` macro.

Reply:
I just made another pass of the doc, and I am taking sam's suggestion to make it very generic. I will take your suggestion for doing the improvement on the code example in the next PR, thanks for it!